src/librustc/infer/error_reporting/mod.rs

```diff
@@ -966,3 +712,515 @@ impl<'tcx> ObligationCause<'tcx> {
+                            .get_opt_name()) {
+                        let name = format!("{}", name);
+                        let path = tcx.item_path_str(def_id);
+                        full_imports.insert(path, name);
+                    }
+                }
+                _ => (),
+            }
+        }
+
+        let mut external_mods = FxHashSet();
+        match item.node {
+            // Importing a fully qualified path, add to type alias table.
+            hir::ItemUse(ref path, hir::UseKind::Single) => {
+                debug!("visit_item itemuse single {:?} {:?}", path, path.def);
+                let ty = self.tcx.item_path_str(path.def.def_id());
+                self.full_imports.insert(format!("{}", ty), format!("{}", item.name));
+            }
+            // Importing everything under a path, inspect recursively all the public exports from
+            // that module and add all of them to the type alias table.
+            hir::ItemUse(ref path, hir::UseKind::Glob) => {
+                debug!("visit_item itemuse glob {:?} {:?}", path, path.def);
+
+                if let Mod(def_id) = path.def {
+                    if let Some(node_id) = self.tcx.hir.as_local_node_id(def_id) {
+                        // Same crate import.
+
+                        // Inspect the module being imported for all visible items and add them to
+                        // the type alias table.
+                        if let hir::map::NodeItem(item) = self.tcx.hir.get(node_id) {
+                            if let hir::ItemMod(ref mod_) = item.node {
+                                // Look at all the items in the imported mod.
+                                for item_id in &mod_.item_ids {
+                                    if let hir::map::NodeItem(item) = self.tcx.hir.get(item_id.id) {
+                                        // We only want to add items that have been exported to the
+                                        // type alias table, check visibility.
+                                        match item.vis {
+                                            hir::Visibility::Public |
+                                            hir::Visibility::Crate => {
+                                                let ty = self.tcx.node_path_str(item.id);
+
+                                                self.full_imports.insert(format!("{}", ty),
+                                                                         format!("{}", item.name));
+                                            }
+                                            _ => (),  // item not visible, don't alias
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    } else {
+                        // Importing from a different crate.
+                        for child in self.tcx.cstore().item_children_untracked(def_id,
```

Comment:
Ah so the "untracked" here means that it's not tracked by incremental compilation, so this is something that we're going to want to handle as this is all happening after the tcx is created. If you need help turning this into a query though just let me know!

Reply:
Could you point me in the right direction of a query being performed? I haven't worked in that part of the compiler yet and would appreciate having a thread to pull from to unravel that sweater.

Reply:
Sure! You can check out https://github.com/rust-lang/rust/issues/41417 for some more info